OneCommunityGlobal/HighestGoodNetworkApp
# Xiaow task system improvement

# Description
This PR aims to cooperate with backend to overhaul the task system logic and improve UI. Major changes include but not limited to: overhaul Task component and task controller display logic, improve logic of task import, addition, move, delete and edit,  improve task reducer logic, and make code more organized.

## Related PRS :
This frontend PR is related to the [#455](https://github.com/OneCommunityGlobal/HGNRest/pull/455#issue-1816682991) backend PR.

## Main changes explained:
- Frequent DOM manipulations are not recommended for React app, so one major change for this PR is that it converts most DOM manipulation to the way react handles its states and make the render output solely controlled by React.
- Task components now have nested lower level Task components, which will only rendered on page when unfolded a higher level task
- "All" filter on the top only works as a filter just like the rest of the filters, and "Fold/Unfold" button is added to control all tasks collapse
- Task controller is moved out of Task as a different component, and it makes sure only one controller is shown on page, and folding higher level task will close the controller for lower level tasks.
- For each task item, parentIds and mother props are revised, and the backend logic for tasks is rewrite to reconcile with this change. It makes sure that the tasks of different levels are properly organized.

- For add task functionality:
![add task](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/77769556/8057040b-72e9-425e-9c0f-ef38c68ac9c8)
  - Now have a correct num (WBS #) showing on the top of the modal, and it depends on where the new task is added. Previously it will not show the correct num for some edge cases.
  - For "status", according to the xlsx file there should be four options: Active, Not Started, Paused and Complete. Proper changes are made to accommodate this. For Task component, different icons are used as well (from top to bottem are Active, Not Started, Paused and Complete)
![status icons](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/77769556/ad2d5f05-d732-4846-b8f6-c0a59410dc64)
  - Now when you add a new task, app will check with backend to make sure database is not changed since you load the page, in a special case which another admin use add a new task to the same WBS, app will alert current user that database has changed, and once user clicks OK, app will refresh the page with the most recent data. This will prevent different users create task with the same number.
![repeated task num created](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/77769556/a02c70a0-51c5-45de-ace2-9af48db7a4ed)
  - Copy functionality should also work now, be sure to copy a task in its controller bar and then use the paste button at the top of the add task modal.
- For import task functionality:
  - now in xlxs file, resources column can have multiple names, but it must be inside of the name list of current project member, otherwise app will warn user.
![import warn](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/77769556/c8043fd8-31c6-4da4-a775-2e9ac2de476d)
  - a reset button is added 

- For move task functionality:
  - now not only the target task will be moved, any child tasks inside of target task will move along with it.
  before move:
![before move](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/77769556/5d2b47c6-594f-4da2-ac25-40ae94d23d7b)
  after move:
![after move](https://github.com/OneCommunityGlobal/HighestGoodNetworkApp/assets/77769556/3d368d4f-356a-407d-b1bc-7ed54e9e0120)

- some other QoL changes, such as tooltips and proper response/loading icon as feedback for any task function operation 
- I disabled 'Level -1' test cases since I replace it with case: EMPTY_TASK_ITEMS and this test is not applicable anymore.

## How to test:
1. check into current branch and backend PR [#455](https://github.com/OneCommunityGlobal/HGNRest/pull/455#issue-1816682991)
2. do `npm install` and `npm run start:local` to run this PR locally
3. log as admin user
4. go to dashboard → Other Links → Projects → choose one Project  → choose one WBS 
5. verify import task (only from empty WBS), add task (both from the top of the table and from task controller), edit, remove, move, copy (then paste to new task) all work as intended.


## Note:
- import task can only be tested in empty WBS, and you could ask Jae for the template xlsx file for testing.
- since a lot of changes were made to accommodate the new task system, it is recommended to create a new WBS and create new tasks for testing move functionality
- to test add task with same number, use two browser pages and login the same WBS, create one task on one page and then try to create a new task on the second page, see if you could get the alert.
- I haven't figure out how the Hours works yet, since you could add a sub task that has way more hours than its mother task, but from the import logic in the backend it seems that the hours of a mother task should be the sum of all of its child tasks, which baffles me
